### PR TITLE
Fix CPU bitwise shifts for out-of-limit shift values

### LINKED ISSUE #70904: Bitwise shifts are broken in some cases if shift is greater than bit size
Left- or right-shifting the bits of an integer tensor by equal or more than the number of bits of the dtype, should always result in `0`. This is not the case for some shifts independent of the `dtype`:

```python
import torch

dtype = torch.uint8
info = torch.iinfo(dtype)
input = torch.tensor(info.max, dtype=dtype)

for other in range(info.bits, 100):
    other = torch.full_like(input, other)

    result_left = input << other
    if result_left != 0:
        print(f"{int(input)} << {int(other)}: {int(result_left)}")

    result_right = input >> other
    if result_right != 0:
        print(f"{int(input)} >> {int(other)}: {int(result_right)}")
```

```
255 << 32: 255
255 >> 32: 255
255 << 33: 254
255 >> 33: 127
255 << 34: 252
255 >> 34: 63
255 << 35: 248
255 >> 35: 31
255 << 36: 240
255 >> 36: 15
255 << 37: 224
255 >> 37: 7
255 << 38: 192
255 >> 38: 3
255 << 39: 128
255 >> 39: 1
255 << 64: 255
255 >> 64: 255
255 << 65: 254
255 >> 65: 127
255 << 66: 252
255 >> 66: 63
255 << 67: 248
255 >> 67: 31
255 << 68: 240
255 >> 68: 15
255 << 69: 224
255 >> 69: 7
255 << 70: 192
255 >> 70: 3
255 << 71: 128
255 >> 71: 1
255 << 96: 255
255 >> 96: 255
255 << 97: 254
255 >> 97: 127
255 << 98: 252
255 >> 98: 63
255 << 99: 248
255 >> 99: 31
```

### PR DESCRIPTION
**Pull request by bjhargrave**
Negative shift values and positive shift values greater than the bit size of the dtype (limit `0..bits`) now yield expected results which are consistent with numpy.

Left shift with an out-of-limit shift value result in a value of `0`. Right shift with an out-of-limit shift value results in a value of `-1` for negative inputs and `0` for non-negative inputs (sign preserving).

Fixes https://github.com/pytorch/pytorch/issues/70904

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Review by ngimel — APPROVED** [review]

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp [hunk 1]
```diff
@@ -314,12 +314,21 @@ void bitwise_xor_kernel(TensorIteratorBase& iter) {
...
+            return static_cast<std::make_unsigned_t<scalar_t>>(a) << b;
+          }
+          return 0;
         },
-        [](Vectorized<scalar_t> a, Vectorized<scalar_t> b) {
-            return a << b;
+        [&zero_vec, &shift_vec](vec_t a, vec_t b) -> vec_t {
+          auto mask = (b >= zero_vec) & (b < shift_vec);
+          a = vec_t::blendv(zero_vec, a, mask);
+          return a << b;
```
Comment (jgong5) [inline aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:331]: May I know what is the problem for vectorized implementation here? The `_mm256_srav_epi32` and `_mm256_sllv_epi32` etc. should already do the right things and we don't have to do extra masking here, right?

Comment (albanD) [reply aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:331]: What do you mean by "the right thing"? These are UB from the standard iirc.

**Review by albanD — APPROVED** [review]
SGTM

**Review by zou3519 — APPROVED** [review]

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp (hunk 1 above)
Comment (jgong5) [reply aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:331]: I mean the semantics of the intrinsics/instructions already guarantee the expected behavior, e.g., when shifting left with outside-range count, zero is returned and when shifting right, sign flag is preserved, no need to do line 329-330 (and similarly for right shift). Or did I miss anything here?

Comment (jgong5) [reply aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:331]: https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#text=_mm256_srav_epi32
https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#text=_mm256_sllv_epi32

Comment (albanD) [reply aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:331]: Thanks for the details. Is tha ttrue for all CPU though? Including older Intel hardware that might implement an older version of the intrinsics? And non-Intel hardware?

Comment (bjhargrave) [reply aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:331]: The nightly macos build is compiled w/o AVX2 and it fails the new test. So a fix is necessary for other CPUs.

An alternate solution, which assumes the SIMD implementations of `operator<<` and `operator>>` are correct, would be to fix the `operator<<` and `operator>>` implementations in `vec_base.h` for when SIMD implementation is not supported. I will try this.

Comment (bjhargrave) [reply aten/src/ATen/native/cpu/BinaryOpsKernel.cpp]: Turns out there was a bug in `vec256` for the [`operator>>` implementation for `int64`](https://github.com/pytorch/pytorch/blob/ba4fb9b6ad0b3a06ae1b15a170cea6f4223bda79/aten/src/ATen/cpu/vec/vec256/vec256_int.h#L1482-L1497). Since there is no shift right arithmetic vector instruction, the implementation emulated the instruction. However the implementation did not work if the shift value was out-of-limit (<0 or >64). When the shift value was out-of-limit, subtracting the shift value from 64 to compute the sign shift will result in a bad sign shift value and bad sign extension value for negative input.

I have added a new commit to the PR which fixes this.

Comment (jgong5) [reply aten/src/ATen/native/cpu/BinaryOpsKernel.cpp]: > Thanks for the details. Is tha ttrue for all CPU though? Including older Intel hardware that might implement an older version of the intrinsics? And non-Intel hardware?

The behavior should be consistent on x86 CPUs in general. In general, I think it is better to address the problem in the vectorized implementation to avoid unnecessary guards where possible.

**Review by jgong5 — APPROVED** [review]

**Review by albanD — COMMENTED** [review]
Thanks everyone!

@pytorchbot merge

**Review by malfet — APPROVED** [review]
